docs/cip/CIP_x_Metadata_Indexes.md

```diff
@@ -0,0 +1,133 @@
+# CIP-X Metadata Indexes
+
+## WARNING: Challenges to discuss
+
+- One of the main challenges with the below proposal is that indices are not per collection and defining them at
+  collection level feels counter-intuitive.
+- Another challenge is that string_value also contains the actual documents which means that the index will be very
+  large and will not be very efficient.
+
+## Status
+
+Current Status: `Under Discussion`
+
+## Motivation
+
+Currently, some users are experiencing extreme slowness when querying with metadata filters, more frequently occurring
+for
+large datasets. This behaviour is due to the full table scan that metadata filters (WHERE clauses) must perform to
+locate the relevant documents.
+
+## Public Interfaces
+
+The proposal suggests the introduction of the following public interfaces:
+
+- `Collection.modify(indices = Index(name='<name>',columns={'col'},operation=ADD|DROP|UPDATE|REBUILD))` - Allows users
+  to
+  define indices on the collection.
+- `Collection` - introduces a new property `indices` that returns a list of all defined indices on the collection.
+
+## Proposed Changes
+
+We propose the introduction of a new function parameter to `Collection.modify` (`indices`) that will allow users to
+define indices on the collection. The indices will be defined list of Index objects. The indices will be created on the
+SQLite database level.
+
+Index Object:
+
+```python
+AllowedColumns = Literal["string_value", "int_value", "float_value", "key"]
+
+
+class Index(BaseModel):
+    name: str
+    columns: Optional[Set[AllowedColumns]] = None
+    operation: Optional[CollectionIndexOperation] = CollectionIndexOperation.ADD
+```
+
+Where `CollectionIndexOperation` is an enum with the following values:
```

Comment:
This feels a bit odd (the operations being part of the index object) and un-pythonic

What about something more like

collection.indices.add
collection.indices[ACCESSOR].update/rebuild
collection.indices.drop(ACCESSOR)

?

Reply:
I agree, but that was the one of the options if we wanted to have `Collection.modify()` if we move out indices as props or attributes of the Collection then the above makes a lot more sense and is much more elegant solution.